cli/register/register.go

```diff
@@ -275,49 +288,88 @@ func DoIt(org, pattern, nodeIdTok, userPw, email, inputFile string, nodeOrgFromF
+			attr = api.NewAttribute("UserInputAttributes", "service", false, false, map[string]interface{}{}) // we reuse this for each service
+			emptyStr := ""
+			service := api.Service{Name: &emptyStr} // we reuse this too
+			if len(inputFileStruct.Services) > 0 {
+				msgPrinter.Printf("Setting service variables...")
+				msgPrinter.Println()
+			}
+			for _, m := range inputFileStruct.Services {
+				service.Org = &m.Org
+				service.Url = &m.Url
+				service.VersionRange = &m.VersionRange
+				attr.Mappings = &m.Variables
+				attrSlice := []api.Attribute{*attr}
+				service.Attributes = &attrSlice
+				httpCode, respBody := cliutils.HorizonPutPost(http.MethodPost, "service/config", []int{201, 200, 400}, service)
+				if httpCode == 400 {
+					if matches := parseRegisterInputError(respBody); matches != nil && len(matches) > 2 {
+						cliutils.Fatal(cliutils.CLI_INPUT_ERROR, msgPrinter.Sprintf("Registration failed because %v Please update the services section in the input file %v. Run 'hzn unregister' and then 'hzn register...' again", matches[0], inputFile))
+					}
+					cliutils.Fatal(cliutils.CLI_INPUT_ERROR, "Error setting service variables from user input file: %v", respBody)
+				}
+			}
+		} else {
+			msgPrinter.Printf("usePolicyInputFormat: %t", usePolicyInputFormat)
 			msgPrinter.Println()
-		}
-		for _, m := range inputFileStruct.Services {
-			service.Org = &m.Org
-			service.Url = &m.Url
-			service.VersionRange = &m.VersionRange
-			attr.Mappings = &m.Variables
-			attrSlice := []api.Attribute{*attr}
-			service.Attributes = &attrSlice
-			httpCode, respBody := cliutils.HorizonPutPost(http.MethodPost, "service/config", []int{201, 200, 400}, service)
-			if httpCode == 400 {
-				if matches := parseRegisterInputError(respBody); matches != nil && len(matches) > 2 {
-					cliutils.Fatal(cliutils.CLI_INPUT_ERROR, msgPrinter.Sprintf("Registration failed because %v Please update the services section in the input file %v. Run 'hzn unregister' and then 'hzn register...' again", matches[0], inputFile))
+			// use policy.UserInput struct
+			attr := api.NewAttribute("UserInputAttributes", "service", false, false, map[string]interface{}{}) // we reuse this for each service
+			emptyStr := ""
+			service := api.Service{Name: &emptyStr} // we reuse this too
+			if len(policyInputFileList) > 0 {
+				msgPrinter.Printf("Setting service variables...")
+				msgPrinter.Println()
+			}
+
+			uiMap := make(map[string]interface{})
+			for _, ui := range policyInputFileList {
+				service.Org = &ui.ServiceOrgid
+				service.Url = &ui.ServiceUrl
+				service.VersionRange = &ui.ServiceVersionRange
+				for _, input := range ui.Inputs {
+					if input.Name != "" {
+						uiMap[input.Name] = input.Value
+					}
+				}
+				attr.Mappings = &uiMap
+				attrSlice := []api.Attribute{*attr}
+				service.Attributes = &attrSlice
+				httpCode, respBody := cliutils.HorizonPutPost(http.MethodPost, "service/config", []int{201, 200, 400}, service)
```

Comment:
For this case, you can just call 'node/userinput' API and pass the whole thing to it. 
Please see https://github.com/open-horizon/anax/blob/master/cli/userinput/userinput.go#L64

Reply:
👍